### PR TITLE
feat(vscode): add proxy support for vscode

### PR DESCRIPTION
**Pull request by eryue0220**
* This pr is to solve #2893

### CODE REVIEW

File: clients/tabby-agent/src/lsp/protocol.ts [hunk 1]
```diff
@@ -177,6 +177,11 @@ export namespace ConfigurationRequest {
  * [Tabby] Defines the config supported to be changed on the client side (IDE).
  */
 export type ClientProvidedConfig = {
+  proxy?: {
+    authorization: string;
+    url: string;
+    noProxy: string | string[];
```
Comment (wsxiaoys) [inline clients/tabby-agent/src/lsp/protocol.ts:183]:
```suggestion
```

I recommend removing this option for convenient of implementation in first step

Comment (eryue0220) [reply clients/tabby-agent/src/lsp/protocol.ts]: Updated.

**Comment by wsxiaoys** [comment]
Please also rebase against main branch

File: clients/tabby-agent/src/TabbyAgent.ts [hunk 2]
```diff
@@ -109,9 +109,12 @@ export class TabbyAgent extends EventEmitter implements Agent {
 
     await loadTlsCaCerts(this.config.tls);
 
+    let hasUpdate = false;
+
```
[the remark this answers is not in the document]
Comment (eryue0220) [reply clients/tabby-agent/src/TabbyAgent.ts]: Updated.

File: clients/tabby-agent/src/lsp/Server.ts [hunk 3]
```diff
@@ -363,6 +363,31 @@ export class Server {
         this.agent.clearConfig("anonymousUsageTracking.disable");
       }
     }
+    if (
```
[the remark this answers is not in the document]
Comment (eryue0220) [reply clients/tabby-agent/src/lsp/Server.ts:379]: Update. I've refactor this function.

[the remark this answers is not in the document]
Comment (wsxiaoys) [reply clients/tabby-agent/src/lsp/Server.ts:379]: `updateConfiguration` still requires refactoring?

[the remark this answers is not in the document]
Comment (eryue0220) [reply clients/tabby-agent/src/lsp/Server.ts]: Ohhhhh, Sorry I forgot that. 

I'll update it later.

[the remark this answers is not in the document]
Comment (eryue0220) [reply clients/tabby-agent/src/lsp/Server.ts]: Done.

File: clients/tabby-agent/src/lsp/Server.ts [hunk 4]
```diff
@@ -344,54 +344,97 @@ export class Server {
...
-        this.agent.updateConfig("anonymousUsageTracking.disable", true);
-      } else {
-        this.agent.clearConfig("anonymousUsageTracking.disable");
+
+    if (!clientProvidedConfig) return;
+
+    const fieldsToCheck = [
+      {
+        key: "server.endpoint",
+        validation: (key: string, x: string) => {
```
Comment (wsxiaoys) [inline clients/tabby-agent/src/lsp/Server.ts:367]:
```suggestion
        validation: (key: keyof ClientProvidedConfig, x: string) => {
```

Just use explict type?

File: clients/tabby-agent/src/lsp/Server.ts [hunk 5]
```diff
@@ -344,54 +344,97 @@ export class Server {
...
+          const prop = clientProvidedConfig?.[key as keyof ClientProvidedConfig];
+          return (
+            prop?.[x as keyof typeof prop] !== undefined && (prop?.[x as keyof typeof prop] as string).trim().length > 0
+          );
+        },
+      },
+      {
+        key: "server.token",
+        validation: (key: string, x: string) => {
```
Comment (wsxiaoys) [inline clients/tabby-agent/src/lsp/Server.ts:376]: ditto, and the function can be reused?

**Review by wsxiaoys — COMMENTED** [review]
Otherwise LGTM

File: clients/tabby-agent/src/lsp/Server.ts (hunk 4 above)
Comment (eryue0220) [reply clients/tabby-agent/src/lsp/Server.ts]: Updated.

File: clients/tabby-agent/src/lsp/Server.ts (hunk 5 above)
Comment (eryue0220) [reply clients/tabby-agent/src/lsp/Server.ts]: Right, I just struggled with type and ignored.

File: clients/tabby-agent/src/AgentConfig.ts [hunk 6]
```diff
@@ -12,6 +12,8 @@ export type AgentConfig = {
     requestTimeout: number;
   };
   proxy: {
+    enabled: boolean;
```
Comment (icycodes) [inline clients/tabby-agent/src/AgentConfig.ts:15]: I prefer to not add `proxy.enabled`. You can set the `proxy.url` to empty to disable the proxy.

Comment (eryue0220) [reply clients/tabby-agent/src/AgentConfig.ts]: Updated.

**Review by wsxiaoys — APPROVED** [review]

File: clients/tabby-agent/src/lsp/Server.ts [hunk 7]
```diff
@@ -344,54 +344,93 @@ export class Server {
...
+          clientProvidedConfig,
+          first as keyof ClientProvidedConfig,
+          second as keyof ClientProvidedConfig,
+        )
+      ) {
+        if (validation(first as keyof ClientProvidedConfig, second!)) {
+          const firstProp = clientProvidedConfig[first as keyof typeof clientProvidedConfig];
+          this.agent.updateConfig(`${first}.${second}`, firstProp?.[second as keyof typeof firstProp]);
+        } else {
+          this.agent.clearConfig(`${first}.${name}`);
```
Comment (icycodes) [inline clients/tabby-agent/src/lsp/Server.ts:413]:
```suggestion
          this.agent.clearConfig(`${first}.${second}`);
```

**Review by icycodes — CHANGES_REQUESTED** [review]

File: clients/tabby-agent/src/lsp/Server.ts (hunk 7 above)
Comment (icycodes) [reply clients/tabby-agent/src/lsp/Server.ts:413]: Otherwise LGTM

Comment (eryue0220) [reply clients/tabby-agent/src/lsp/Server.ts]: Updated.

Thanks you.

**Review by icycodes — APPROVED** [review]

**Pull request merged**